### ISSUE TITLE
stringToUtf8ByteArray produces illegal UTF-8 for non-BMP characters

### ISSUE THREAD

**Issue by lokedhs** [google/closure-library#470]
When using `stringToUtf8ByteArray` to convert a string into a sequence of UTF-8 encoded bytes, any characters outside of the BMP (i.e. characters with a code point > 0xFFFF) gets encoded as a sequence of two surrogate pairs.

Encoding surrogate pairs as UTF-8 is illegal and will cause a proper decoder to fail. See RFC 3629 which states:

_"The definition of UTF-8 prohibits encoding character numbers between U+D800 and U+DFFF, which are reserved for use with the UTF-16 encoding form (as surrogate pairs) and do not directly represent characters."_

**Comment by fbsolo** [comment]
I started working on this and I hit a wall because of issues with the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/charCodeAt">charCodeAt</a> function. In some email traffic, Lokedhs explained that the solution would involve extracting the high and low surrogate pairs that map to the "special" characters - those that have <a href="http://en.wikipedia.org/wiki/UTF-8">code points</a> greater than 0xFFFF. The charCodeAt function would handle all this. I had trouble getting it to work and I finally built a small test file in my local Closure instance for development, testing in Chrome. No luck. Then I built this <a href="https://jsfiddle.net/jsyvqhec/">JSFiddle</a> which does this:

![test_fiddlejs_code_in_jsfiddle](https://cloud.githubusercontent.com/assets/5409877/7824748/60b2c604-03ba-11e5-83fb-8ef16c61befc.png)

I placed the code from that JSFiddle back in the local Closure test instance and in Chrome, it does this:

![test_fiddlejs_code_in_chrome](https://cloud.githubusercontent.com/assets/5409877/7824754/70388a78-03ba-11e5-89da-030084699a8a.png)

In IE, it does this:

![test_fiddlejs_code_in_ie](https://cloud.githubusercontent.com/assets/5409877/7824756/779c7e3c-03ba-11e5-8e5d-8af8330945f4.png)

I found the "overload" for charCodeAt <a href="http://stackoverflow.com/questions/7123140/javascript-apply-or-call-used-on-charcodeat">here</a>.

Bottom line: I don't see a reliable way to grab the surrogate pairs for characters outside the BMP. Until I figure this out, or someone has a suggestion, I can't proceed. I almost think I discovered a more fundamental bug.

**Comment by lokedhs** [comment]
I wouldn't use that one at all. I would rather simply iterate over the characters and do surrogate processing manually. Something like this (note, completely untested):

```
var i;
var pendingHighSurrogate = null;
var result = [];
for(i = 0 ; i < s.length ; i++) {
    var ch = s[i];
    // High surrogate?
    if(ch >= 0xD800 && ch <= 0xDBFF) {
        if(pendingHighSurrogate != null) {
            throw "Illegal format: Two high surrogates in a row";
        }
        pendingHighSurrogate = ch;
    }
    else if(ch >= 0xDC00 && ch <= 0xDFFF) {
        if(pendingHighSurrogate == null) {
            throw "Illegal format: Low surrogate not preceeded by high surrogate";
        }
        result.push(encodeCharAsUtf8(((pendingHighSurrogate - 0xD800) << 10) | (c - 0xDC00)));
        pendingHighSurrogate = null;
    }
    else {
        result.push(encodeCharAsUtf8(ch));
    }
}
if(pendingHighSurrogate) {
    throw "Illegal format: truncated surrogate pair";
}
```

**Comment by concavelenz** [comment]
Thanks for the report, this behavior should at least be documented on for the function.

**Comment by shicks** [comment]
This issue is being closed as a preparation step before archiving the repository.  See the [README](https://github.com/google/closure-library) for more detail.
